storage/range.go

```diff
@@ -760,42 +759,79 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
 // struct to get args and reply and then applying the command to the
 // state machine via applyRaftCommand(). The error result is sent on
 // the command's done channel, if available.
-//
-// TODO(spencer): Differentiate between errors caused by the normal culprits --
-// bad inputs from clients, stale information, etc. and errors which might
-// cause the range replicas to diverge -- running out of disk space, underlying
-// rocksdb corruption, etc. Do a careful code audit to make sure we identify
-// errors which should be classified as a ReplicaCorruptionError--when those
-// bubble up to the point where we've just tried to execute a Raft command, the
-// Raft replica would need to stall itself.
 func (r *Range) processRaftCommand(idKey cmdIDKey, index uint64, raftCmd proto.InternalRaftCommand) error {
 	if index == 0 {
 		log.Fatalc(r.context(), "processRaftCommand requires a non-zero index")
 	}
+
 	r.Lock()
 	cmd := r.pendingCmds[idKey]
 	delete(r.pendingCmds, idKey)
 	r.Unlock()
 
 	args := raftCmd.Cmd.GetValue().(proto.Request)
+	origin := proto.RaftNodeID(raftCmd.OriginNodeID)
+	var reply proto.Response
+	var ctx context.Context
+	var err error
 	if cmd != nil {
 		// We initiated this command, so use the caller-supplied reply.
-		err := r.applyRaftCommand(cmd.ctx, index, proto.RaftNodeID(raftCmd.OriginNodeID), args, cmd.Reply)
-		cmd.done <- err
-		return err
+		reply = cmd.Reply
+		ctx = cmd.ctx
+	} else {
+		// This command originated elsewhere so we must create a new reply buffer.
+		reply = args.CreateReply()
+		ctx = r.context()
+	}
+
+	// Verify the leader lease is held; Note that we don't require the leader
+	// lease when trying to grant the leader lease!
+	held, expired := verifyLeaderLease(r.getLease(), origin, args.Header().Timestamp)
+	if (!held || expired) && args.Method() != proto.InternalLeaderLease {
+		// This error case is somewhat special: Any command that makes it into
+		// Raft should have the leader lease held by the proposer. The only
+		// case that violates this assumption is that of the proposer having a
+		// corrupted internal state and erroneously assuming leadership. Then,
```

Comment:
Are you sure? The "proposer" here is the replica which put the command into the Raft black box (see `rng.proposeRaftCommand`), it doesn't matter which replica turns out to be the Raft leader committing it.

I remember now why this is really here though, it's because checking the leader lease once isn't enough: it doesn't remain the same (for a fixed timestamp) as new leader leases are passed out, so a command that spends a long time in Raft may pop out and find it's been proposed by a replica whose leadership has by then ceded to be.
I would argue however that it's fine to execute such commands without the additional check (unless we're worried about corrupted replicas) because a command having been proposed means that the leader lease was held at propose time, and any later lease holder will have a timestamp cache compatible with any such read from a previous lease period. Writes aren't really affected by leases as much, so that leaves admin commands, which are ok too. There might still be some subtleties so I'd leave this in for now.